### DISCUSSION TITLE
Add Tests, Tools and Guidelines for Content Accessibility

### DISCUSSION THREAD

**Discussion by rafaelcastrocouto** [godotengine/godot-proposals#6490] [General Discussions]
### Describe the project you are working on

https://github.com/spicylobstergames/shotcaller-godot
Minimalistic 1v1 MOBA/RTS in top-down pixel style

### Describe the problem or limitation you are having in your project

Having a hard time assuring that I'm providing accessible interfaces.

For eg. the default button focus color is replaced by font hover color and removing the focus style leads to a [failure in accessibility tests](https://www.w3.org/WAI/WCAG21/Techniques/failures/F78). Just to be clear, focus color (and styles) should always override hover colors (and styles) and that's easy to test, and if a developer somehow changes it with scripts we could raise an accessibility warning. Fixing this kind of misbehavior should be a top priority imho. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The current User interface has very limited features that help developers to deploy accessible content.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Content Accessibility Tests, Tools and Guidelines should provide means for making games, applications, and other content accessible to people with disabilities. 

- Tests would allow us to check if the editor itself and developed games are not breaking simple and testable accessibility rules.

- Tools would make easier by enable and validate accessible environments and providing quick shortcuts to set them up. 

- Guidelines are not required but provide insights in order to develop more accessible games.

Here's a very useful list of guidelines that can also be used to generate tools and tests https://www.w3.org/WAI/WCAG21/Techniques/

### If this enhancement will not be used often, can it be worked around with a few lines of script?

N/A

### Is there a reason why this should be core and not an add-on in the asset library?

N/A

**Comment by YuriSizov** [comment]
I'm converting this to a discussion because this is a very general proposal that is not immediately actionable. The idea is valid, but we need more specific descriptions for tools and tests that can be implemented (what problem they solve, how do they solve it, how do users interface with them, etc). As for guidelines I feel like that's not something the project has to offer an opinion. It's a broad subject, applicable to any application or game, not to a specific engine (though official documentation for the aforementioned tests and tools should be created alongside them).

**Reply by erlend-sh** [reply]
This is becoming a mainstream accessibility solution in many Rust game engines (Bevy) and UI frameworks (egui): https://github.com/AccessKit/accesskit

**Reply by bram-dingelstad** [reply]
I might agree with the actionability of this request, but I feel it should be a top priority for the project to provide accessible interfaces. Yes, it differs between different projects, but to have a basis for good accessible UI for players is important if we want Godot to be a used by developers who have accessbility needs, or by players that have these needs.

If you look at the web platform (HTML, etc) it has broad support for accessible interfaces that are generalist enough to cover most if not all use-cases. Having a similar foundation in Godot would be a great selling point for people to start using it over alternatives.

As for the actionability: it would be best to have some consult with interest groups to find out what things are in need of prioritisation when it comes to tools and tests. Guidelines could be simular in the way that we have suggestions on how to do localisation: have good documentation & examples on how to tackle the most common cases for accessibility to make it easier for people to get started on tackling the issue in their games.

I'm definitely willing to contribute to a roadmap/content in order to make Godot more accessible.

**Reply by YuriSizov** [reply]
As I've said, the idea is valid, but it lacks the necessary details for a proposal. Contacting the interested parties and discussing the issue with them is a good idea, and is a path to having a solid and actionable proposal opened.

**Reply by bram-dingelstad** [reply]
Isn't the contacting of interested parties actionable enough for a proposal? I get that the standard for a proposal is to be actionable and concrete, but this seems like a situation in which it would be hard to fit that standard immediately. 🤔

Isn't the direction and idea of this proposal enough to move towards a better state of the project? 😊
If we aren't gonna have this as a proposal, what would you suggest the way to move forward towards a proposal that is actionable & concrete enough? ✨

**Reply by YuriSizov** [reply]
@bram-dingelstad A proposal _issue_ is a start of the implementation part. Contacting stakeholders should be done ahead of time, for any proposal. You can contact them and gather their opinions in any way you like. You can invite them here, to this proposal _discussion_ too.

Currently there is nothing to implement, so we can't have a proposal issue. There is an idea for an improvement, and that's fine, but it needs refinement. As I mentioned in the OP of this thread, a resulting proposal or proposals would answer to the questions such as what problem does it solve, how does it solve it, how do users interface with these enhancements, etc. (hence the proposal template). If you can look at the OP and see where to start implementing the idea, then it's an actionable proposal. If you can look at the OP and say "That's a good opportunity to start a discussion" then it's not an actionable proposal.

I feel like some users think that opening a proposal is like submitting a feature request for the engine maintainers, like this is an idea hub of sorts. This is not how Godot is developed, and this is not what this repository is. We specifically have this split between issues and discussions, so users can use discussions to figure out how to solve a problem and if it affects others. It's an alternative to discussing the same on Discord, or Reddit, or some other private or public platform. The issues here, on the other hand, are a description of something that can be actually worked on by some willing contributor. So it needs to be descriptive and specific. The more technical it is, the better. I understand, that not everyone suggesting an improvement can talk engine code, but at least a high level explanation should be provided with examples and use cases. And the target audience is not the maintainers, but all contributors. So it cannot be a request for the engine team to figure out. It needs to be something that can invite anyone interested to implement it.

I think discussing this meta takes away from the possible discussion of the issue at hand we can have here. So I invite you all to the contributors chat over at https://chat.godotengine.org if you want to better understand the open source workflow of Godot and how the project is managed and why.

**Reply by bram-dingelstad** [reply]
I understand what you're saying and I am not trying to undermine the workflow that has been established in the community for tackling problems. I understand your dichatomy of issues and discussions in this repository and how they're handled.

I however think that there are some issues that can't be resolved within a "technical" perspective, but ask for a ideological or priority from the community. However, putting this in a discussion format makes it a bit too ephemeral I feel. Proposing to make priority for accessibility seems like a worthwhile proposal, even though it doesn't have concrete technical implementation from the getgo.

Is there a way that we can prioritise this group of features as a proposal?

**Reply by rafaelcastrocouto** [reply]
I've updated the problem description with a specific case showing what problem does it solve, how does it solve it and how do users interface with these enhancements, I hope it helps.
I'll also leave the original issue link here for overall context https://github.com/godotengine/godot/issues/74489

**Reply by YuriSizov** [reply]
This is not the platform for rallying the community though. This is a platform that hosts results of such rallying. And if you want to engage the community without a complete feature proposal, you should do this where the users are: on Discord, on Twitter, on Mastodon, on Reddit, on Godot Forums, in local communities.

Having a very visible proposal with tons of upvotes but no technical detail on how it should work creates more noise for contributors than signal. Aye, it shows interest, and aye, there can be interesting discussions which are happening in the comments. But without implementation steps it normally just goes stale, without a resolution but with a huge target for the community to point at and say that we don't care. Which is why making actionable proposals that contributors can actually pick up and put work into is extremely important.

PS. We have like 3k+ proposals, and not all of them fit the description that I'm painting. We started sifting that pile, but 4.0 took priority. The end goal would be to reduce that number dramatically by enforcing some standards and moving some to discussions while closing others. In the end the issues tab should represent a soft roadmap for contributors.

**Reply by erlend-sh** [reply]
> This is not the platform for rallying the community though.

I’m a bit confused by this. Earlier you seem to be saying that this Discussions forum is an alternative to those other channels:

> We specifically have this split between issues and discussions, so users can use discussions to figure out how to solve a problem and if it affects others. It's an alternative to discussing the same on Discord, or Reddit, or some other private or public platform.

I get that that Issues aren’t yet the right place for this, but are you saying this shouldn’t exist in Discussions as well now? The Godot users are spread across many discussion platforms, and this is one of them. The only non-chatty alternative among them is https://godotforums.org/, which is effectively the same as this GitHub board except less official and one step further removed from the submission process.

Just making sure we are at least on the same page about whether or not the discussion as it’s happening now is as intended ☺️

**Reply by YuriSizov** [reply]
@erlend-sh I was replying to the question about putting this into a proposal. Discussions are fine. Otherwise I wouldn't move this ticket here, I would've closed it.

**Reply by akien-mga** [reply]
Just to clarify, the actionability ("Do we have a clear design we can implement right now for Godot 4.1?") isn't tied to how important the proposal is to the roadmap.

I fully agree that more accessibility, both for Godot itself and for games made with it, is a very important goal and something we should work towards. There's e.g. some work being done by @bruvzg and others on implementing AccessKit (see https://github.com/godotengine/godot/issues/58074).

I suggest not getting too dragged down in meta discussions on the workflow (i.e. this subthread) and use the handy thread replies feature from this Discussion to discuss the multiple relevant topics that this proposal opens up. Once we do reach one or more clear ideas for actionable items to implement in Godot, any interested contributor to this discussion can summarize the key points and to-do's as one or more technical implementation proposals (i.e. Issues in this repo), and we can put it on the roadmap (ideally with an identified volunteer to do the implementation).

**Comment by QueenOfSquiggles** [comment]
**TL;DR** : I am wholly in favour of implementing accessibility features into the engine by default, and think that would be a boon for many developers and players alike. However, in the short term I feel that addons and GDExtension will allow for greater flexibility in development as well as allowing users to create bespoke solutions that integrate seamlessly with their specific game.

I really like this proposal/discussion. I think focusing on accessibility features by default will make the process of creating accessible games much easier. I personally try to create accessible features in my games, but I am definitely limited to implementations that I am aware of. (i.e. reducing/removing flashing lights, subtitles, TTS, and such). I would love to see as many accessibility features as possible be integrated directly into the Godot Engine if nothing more than to make the process that much easier and enjoyable. However I do agree with Yuri that not everything is "immediately actionable". And arguably more regular participants probably have a bigger say than me.

I do believe that in the short term, implementations of accessibility features can be created through community developed addons and such. With GDExtension in Godot 4 I'm expecting that integrating native libraries will be much easier. It will encourage development of "extensions"(?) for hooking into accessibility systems such as OS-side TTS (Systems that blind users are likely to have installed already, not Godot's existing TTS implementation). And I imagine many more features could be added through the use of addons in the future. From what I've experienced, this is more or less the default for the Godot Community. That is, if there's a feature you like, first see if it can be implemented with an addon. If not, try to find a way that it can be implemented, including changing the base code of the engine. Before I go too far, I am not assuming you don't know this, I'm including it for anyone who might be reading this discussion in the future, who is less familiar with this process.

This engine's modularity in development lets individuals such as myself to create addons to perfectly fit my needs and/or the needs of my players. I actually made a subtitle addon for handling 3D spatial audio in a somewhat reasonable way. I don't think I've updated it past 3.2. And honestly I really would appreciate an engine-default subtitling system, I would probably use it for all of my games if it were added. But making it as an addon allowed me to creature a bespoke solution to my project, a game using 3D spatial audio, that would render it unusable to another project such as any 2D game.

And just let me re-iterate, I am completely in support of an engine default solution, because that would enable developers to save time as well as enable the engine itself to become more accessible. I feel this could be more actionable if we discussed very specific features and implementations, the kinds of details that could be made into an issue/feature bounty. 

And it's entirely possible I'm completely missing the point. I was having a hard time parsing the links provided. That might just be me having a hard time processing that information load.

**Reply by Calinou** [reply]
> It will encourage development of "extensions"(?) for hooking into accessibility systems such as OS-side TTS (Systems that blind users are likely to have installed already, not Godot's existing TTS implementation)

Godot already uses OS-side TTS APIs, but you're referring to [screen reader support](https://github.com/godotengine/godot/issues/58074) here. These are entirely different APIs with different use cases, and are meant to coexist in parallel.

**Reply by Zireael07** [reply]
As @Calinou said, TTS is NOT screen reader. Pinging @ndarilek here since they are the expert, though

**Reply by bruvzg** [reply]
> I do believe that in the short term, implementations of accessibility features can be created through community developed addons and such. With GDExtension in Godot 4 I'm expecting that integrating native libraries will be much easier. It will encourage development of "extensions"(?) for hooking into accessibility systems such as OS-side

For the reference, here's a proof of concept OS screen reader support GDExtension demo (uses AccessKit, and works on all desktop platforms) - https://github.com/bruvzg/godot_accesskit_demo, some engine changes are still necessary (https://github.com/godotengine/godot/pull/72886) since screen reader API is really sensitive to initialization time (should be done immediately after window creation, or OS will miss it). But otherwise, GDExtension should be able to collect most of the required metadata from the Godot nodes (some complex UI elements like Tree, ItemList and RTL might need some modifications to expose internal structure).

**Comment by rafaelcastrocouto** [comment]
Here's how Visual Studio is Integrating an Accessibility Checker
https://devblogs.microsoft.com/visualstudio/developing-accessible-apps-with-visual-studio/

**Reply by rafaelcastrocouto** [reply]
And a here's a fresh Github article highlighting a developer struggle and how accessibility is lacking and most devices and systems are designed by fully-abled people: [From gaming with your eyes to coding with AI: New frontiers for accessibility](https://github.com/readme/featured/open-source-accessibility)
